src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenDeconstructTests.cs

```diff
@@ -0,0 +1,521 @@
+                // (8,18): error CS8208: The Deconstruct method for type 'C' must have only out parameters.
+                //         (x, y) = new C();
+                Diagnostic(ErrorCode.ERR_DeconstructRequiresOutParams, "new C()").WithArguments("C").WithLocation(8, 18)
+                );
+        }
+
+        [Fact]
+        public void DeconstructCanHaveReturnType()
+        {
+            string source = @"
+class C
+{
+    static void Main()
+    {
+        long x;
+        string y;
+
+        (x, y) = new C();
+        System.Console.WriteLine(x + "" "" + y);
+    }
+
+    public int Deconstruct(out int a, out string b)
+    {
+        a = 1;
+        b = ""hello"";
+        return 42;
+    }
+}
+";
+
+            var comp = CompileAndVerify(source, expectedOutput: "1 hello", parseOptions: TestOptions.Regular.WithTuplesFeature());
+            comp.VerifyDiagnostics();
+        }
+
+        [Fact]
+        public void DeconstructDataFlow()
+        {
+            string source = @"
+class C
+{
+    static void Main()
+    {
+        int x, y;
+        (x, y) = new C(); // x and y are assigned here, so no complaints on usage of un-initialized locals on the line below
+        System.Console.WriteLine(x + "" "" + y);
+    }
+
+    public void Deconstruct(out int a, out int b)
+    {
+        a = 1;
+        b = 2;
+    }
+}
+";
+            var comp = CreateCompilationWithMscorlib(source, parseOptions: TestOptions.Regular.WithTuplesFeature());
+            comp.VerifyDiagnostics();
+        }
+
+        [Fact]
+        public void VerifyExecutionOrder()
+        {
+            string source = @"
+using System;
+class C
+{
+    int x { set { Console.WriteLine($""setX""); } }
+    int y { set { Console.WriteLine($""setY""); } }
+
+    C getHolderForX() { Console.WriteLine(""getHolderforX""); return this; }
+    C getHolderForY() { Console.WriteLine(""getHolderforY""); return this; }
+    C getDeconstructReceiver() { Console.WriteLine(""getDeconstructReceiver""); return this; }
+
+    static void Main()
+    {
+        C c = new C();
+        int z; // PROTOTYPE(tuples) this should be removed once the return-type issue is fixed
+        (c.getHolderForX().x, c.getHolderForY().y, z) = c.getDeconstructReceiver();
+    }
+    public void Deconstruct(out D1 x, out D2 y, out int z) { x = new D1(); y = new D2(); z = 3; Console.WriteLine(""Deconstruct""); }
+}
+class D1
+{
+    public static implicit operator int(D1 d) { Console.WriteLine(""Conversion1""); return 1; }
+}
+class D2
+{
+    public static implicit operator int(D2 d) { Console.WriteLine(""Conversion2""); return 2; }
+}
+";
+
+            string expected =
+@"getHolderforX
+getHolderforY
+getDeconstructReceiver
+Deconstruct
+Conversion1
+setX
+Conversion2
+setY
+";
+            var comp = CompileAndVerify(source, expectedOutput: expected, parseOptions: TestOptions.Regular.WithTuplesFeature());
+            comp.VerifyDiagnostics();
+        }
+
+        [Fact]
```

Comment:
Some tests ideas:
* ref returning methods on the left. Check sideeffects.
* Foo()[Bar()]     on the left, where all Foo, Bar and the indexer have sideeffects - print something
* check IL in simple/common cases
* (.... ) =  async Expr();